lib/grandpa/grandpa.go

```
@@ -1281,3 +1167,209 @@ func (s *Service) lenVotes(stage Subround) int {
+	if err != nil {
+		return fmt.Errorf("compacting justification: %w", err)
+	}
+
+	err = s.grandpaState.SetPrecommits(commitMessage.Round, commitMessage.SetID, preCommitSigned)
+	if err != nil {
+		return fmt.Errorf("setting precommits: %w", err)
+	}
+
+	// TODO: re-add catch-up logic (#1531)
+	return nil
+}
+
+func verifyCommitMessageJustification(commitMessage *CommitMessage, setID uint64, threshold uint64,
+	authorities []*types.Authority, blockState BlockState) error {
+	if len(commitMessage.Precommits) != len(commitMessage.AuthData) {
+		return fmt.Errorf("%w: precommits len: %d, authorities len: %d",
+			ErrPrecommitSignatureMismatch, len(commitMessage.Precommits), len(commitMessage.AuthData))
+	}
+
+	if commitMessage.SetID != setID {
+		return fmt.Errorf("%w: grandpa state set id %d, set id in the commit message %d",
+			ErrSetIDMismatch, setID, commitMessage.SetID)
+	}
+
+	highestFinalizedHeader, err := blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("getting highest finalised header: %w", err)
+	}
+
+	isDescendant, err := blockState.IsDescendantOf(highestFinalizedHeader.Hash(), commitMessage.Vote.Hash)
+	if err != nil {
+		return fmt.Errorf("verifying ancestry of highest finalised block: %w", err)
+	}
+
+	if !isDescendant {
+		return fmt.Errorf("%w: vote hash %s and highest finalised header %s",
+			errVoteBlockMismatch, commitMessage.Vote.Hash.Short(), highestFinalizedHeader.Hash())
+	}
+
+	eqvVoters := getEquivocatoryVoters(commitMessage.AuthData)
+	var totalValidPrecommits int
+	for i, pc := range commitMessage.Precommits {
+		justification := &SignedVote{
+			Vote:        pc,
+			Signature:   commitMessage.AuthData[i].Signature,
+			AuthorityID: commitMessage.AuthData[i].AuthorityID,
+		}
+
+		err := verifyJustification(justification, commitMessage.Round, setID, precommit, authorities)
+		if err != nil {
+			logger.Errorf("verifying justification: %", err)
+			continue
```

Should'nt we return an error if verification fails for one of the precommit? :thinking:

---

Not really, if we fail to verify some pre-commit we don't count it in the `totalValidPrecommits`, so since we can validate  > 2/3 pre-commits we don't need to return an error